### PR TITLE
Address EKR's comments from pre-17 readthrough. Changes include:

### PR DESCRIPTION
**Pull request by ekr**
- Harmonize requirements for resumption cipher suites:
  With 0-RTT you have to use the whole cipher suite but with
  resumption you only need to match the KDF (which is a crypto
  requirement). This allows you to use PSKs but still negotiate
  new cipher suites.

- Clarify that extensions are generally freshly negotiated in
  resumption.

- Flesh out IANA Considerations.

- Remove some TODO markers, including the one on recommending specific
  groups. If the WG wants to come to consensus on this, I will take
  a PR for it.

### CODE REVIEW

File: draft-ietf-tls-tls13.md [hunk 1]
```diff
@@ -2248,14 +2215,13 @@ the KDF. For externally established PSKs, the Hash algorithm MUST be set when th
...
 Prior to accepting PSK key establishment, the server MUST validate the
-corresponding binder value (see {{psk-binder}} below) If this
-value is not present or does not validate the server MUST abort the
-handshake.  Servers SHOULD NOT attempt to validate multiple binders;
-rather they SHOULD select a single PSK and validate solely the
-binder that corresponds to that PSK.
-
-In order to accept PSK key establishment, the server sends a
-"pre_shared_key" extension indicating the selected identity. 
+corresponding binder value (see {{psk-binder}} below) If this value is
```
Comment (kaduk) [inline draft-ietf-tls-tls13.md]: full stop

File: draft-ietf-tls-tls13.md [hunk 2]
```diff
@@ -2248,14 +2215,13 @@ the KDF. For externally established PSKs, the Hash algorithm MUST be set when th
...
+not present or does not validate the server MUST abort the handshake.
```
Comment (kaduk) [inline draft-ietf-tls-tls13.md]: comma after "validate"?

File: draft-ietf-tls-tls13.md [hunk 3]
```diff
@@ -885,16 +885,17 @@ If the client has not provided a sufficient "key_share" extension (e.g., it
...
          Client                                               Server
 
          ClientHello
-           + key_share             -------->
+         + key_share               -------->
                                    <--------       HelloRetryRequest
+                                                         + key_share
 
          ClientHello
            + key_share             -------->
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: It seems that indentation for this is rather inconsistent. If you're changing it above, I'd suggest making this and others match as well.

File: draft-ietf-tls-tls13.md [hunk 4]
```diff
@@ -1720,43 +1704,34 @@ Here:
...
+an extension which is not specified for a given message it MUST
+abort the handshake with an "illegal_parameter" alert.
+
+The server MUST NOT send any extensions which did not appear in the
+corresponding ClientHello (with the exception of "cookie", as
+described in {{cookie}}). If the client received an extension which
+it did not offer in its ClientHello, it MUST abort the handshake with an
+"unsupported_extension" alert. "server-oriented" extensions are supported by
+having the client send an empty dummy extension indicating support
+for that extension type
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: missing period

File: draft-ietf-tls-tls13.md [hunk 5]
```diff
@@ -1720,43 +1704,34 @@ Here:
...
+and HelloRetryRequest messages. The table in {{iana-considerations}}
+indicates where a given extension may appear. If the client receives
+an extension which is not specified for a given message it MUST
+abort the handshake with an "illegal_parameter" alert.
+
+The server MUST NOT send any extensions which did not appear in the
+corresponding ClientHello (with the exception of "cookie", as
+described in {{cookie}}). If the client received an extension which
+it did not offer in its ClientHello, it MUST abort the handshake with an
+"unsupported_extension" alert. "server-oriented" extensions are supported by
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: Missing capitalization for first word of sentence. I suggest just dropping the quotes and the dash.

File: draft-ietf-tls-tls13.md [hunk 6]
```diff
@@ -2115,23 +2088,20 @@ generated independently.  Clients MUST NOT offer multiple
 KeyShareEntry values for the same group.  Clients MUST NOT offer any
 KeyShareEntry values for groups not listed in the client's
 "supported_groups" extension.  Servers MAY check for violations of
-these rules and and MAY abort the handshake with an
+these rules and and abort the handshake with an
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: double "and"

File: draft-ietf-tls-tls13.md (hunk 1 above)
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: missing period at end of sentence

File: draft-ietf-tls-tls13.md [hunk 7]
```diff
@@ -4477,25 +4420,21 @@ See {{zero-rtt-backwards-compatibility}}.
...
-highest version supported by the server.
-
 Some legacy server implementations are known to not implement the TLS
 specification properly and might abort connections upon encountering
 TLS extensions or versions which it is not aware of. Interoperability
 with buggy servers is a complex topic beyond the scope of this document.
 Multiple connection attempts may be required in order to negotiate
 a backwards compatible connection, however this practice is vulnerable
-to downgrade attacks and is NOT RECOMMENDED.
+to downgrade attacks and is NOT RECOMMENDED
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: accidental removal of a correct period

File: draft-ietf-tls-tls13.md [hunk 8]
```diff
@@ -1720,43 +1704,34 @@ Here:
...
-[[TODO: update this and the previous paragraph to cover PSK-based resumption.]]
+The client sends its extensions in the ClientHello. The server MAY
+send extensions in the ServerHello, EncryptedExtensions, Certificate,
+and HelloRetryRequest messages. The table in {{iana-considerations}}
+indicates where a given extension may appear. If the client receives
+an extension which is not specified for a given message it MUST
+abort the handshake with an "illegal_parameter" alert.
+
+The server MUST NOT send any extensions which did not appear in the
+corresponding ClientHello (with the exception of "cookie", as
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: I suggest dropping the parenthesis here to match the styling of the exception in the first sentence of the next paragraph.

File: draft-ietf-tls-tls13.md [hunk 9]
```diff
@@ -1720,43 +1704,34 @@ Here:
...
-[[TODO: update this and the previous paragraph to cover PSK-based resumption.]]
+The client sends its extensions in the ClientHello. The server MAY
+send extensions in the ServerHello, EncryptedExtensions, Certificate,
+and HelloRetryRequest messages. The table in {{iana-considerations}}
+indicates where a given extension may appear. If the client receives
+an extension which is not specified for a given message it MUST
+abort the handshake with an "illegal_parameter" alert.
+
+The server MUST NOT send any extensions which did not appear in the
+corresponding ClientHello with the exception of the NewSessionTicket
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: nit: comma after "ClientHello"

File: draft-ietf-tls-tls13.md [hunk 10]
```diff
@@ -1720,43 +1704,34 @@ Here:
...
+message and the "cookie" extension in the HelloRetryRequest message).
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: errant closing paren leftover

File: draft-ietf-tls-tls13.md [hunk 11]
```diff
@@ -1720,43 +1704,34 @@ Here:
...
+Upon receiving an unexpected extension, it MUST abort the handshake with an
+"unsupported_extension" alert. Server-oriented extensions are supported by
+having the client send an empty dummy extension indicating support
```
Comment (davegarrett) [inline draft-ietf-tls-tls13.md]: Wording nitpicking: "dummy" doesn't feel like the ideal word here, as it does have meaning. Maybe instead be explicit: "an extension with zero-length extension_data"

**Comment by ekr** [comment]
Updated to the merged version.

**Pull request closed without merging**